### PR TITLE
Added thinning operation - julia v0.7

### PR DESCRIPTION
**Pull request by lucianolorenti**

### CODE REVIEW

**Comment by juliohm** [comment]
Awesome, I am amazed how Travis builds are so much quicker in Julia v0.7 because of the new package system. No need to clone a bunch of git repos anymore.

**Comment by juliohm** [comment]
Hi Luciano, do you see any button on GitHub to give me permission to push commits to your thinning_juliav0.7 branch? I have made some changes to the code that I would like to push for you to review.

If you can find a button on this PR to give me access to your fork of ImageMorphology.jl, let me know.

**Comment by lucianolorenti** [comment]
I have the option 'Allow edits from maintainers' enabled. Is this what you mean?

**Comment by juliohm** [comment]
I don't know, probably. Tried to follow GitHub instruction to update the PR with more commits, but git failed locally. I will investigate it more.

**Comment by juliohm** [comment]
@lucianolorenti apparently you have to give me access to your fork, I am not sure, maybe add me as a contributor there? If I push to your branch in your fork, the PR will be automatically updated here.

I've seen people updating PRs before without asking for permission, I don't know how that works.

**Comment by juliohm** [comment]
Nevermind, found the documentation, reading through it: https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

**Comment by juliohm** [comment]
@lucianolorenti could you please review the new logic in `thinning_impl`? Notice that it is not allocating memory anymore.

The last change I would like to do is perform the padding once outside of the iteration as pre and post-preocessing steps. I tried to do it locally, but for some reason the tests are not passing, will work on it a little more.

**Comment by juliohm** [comment]
Below is my current attempt to pad the image outside of the loop:

```julia
function thinning_impl(img::AbstractArray{Bool,2}, algo::GuoAlgo)
    # pad input image
    prev = falses(size(img).+2)
    prev[2:end-1,2:end-1] = img

    # perform single iteration
    it = 1; curr = copy(prev)
    guo_iteration!(curr, isodd(it))

    # loop if necessary
    while prev != curr
        it += 1; prev .= curr
        guo_iteration!(curr, isodd(it))
    end

    # unpad result
    curr[2:end-1,2:end-1]
end

# perform Guo iteration on already padded image
function guo_iteration!(img::AbstractArray{Bool,2}, odd_iteration::Bool)
    nx, ny = size(img)
    for j=2:ny-1, i=2:nx-1
        img[i,j] || continue
        p1 = img[i-1,j-1]
        p2 = img[i-1,j]
        p3 = img[i-1,j+1]
        p4 = img[i,j+1]
        p5 = img[i+1,j+1]
        p6 = img[i+1,j]
        p7 = img[i+1,j-1]
        p8 = img[i,j-1]
        C = (!p2 && (p3 || p4)) + (!p4 && (p5 || p6)) + (!p6 && (p7 || p8)) + (!p8 && (p1 || p2))
        N1 = (p1 || p2) + (p3 || p4) + (p5 || p6) + (p7 || p8)
        N2 = (p2 || p3) + (p4 || p5) + (p6 || p7) + (p8 || p1)
        N = min(N1, N2)
        if odd_iteration
            O = (p2 || p3 || !p5) && p4
        else
            O = (p6 || p7 || !p1) && p8
        end
        if C == 1 && (2 ≤ N ≤ 3) && !O
            img[i,j] = false
        end
    end
end
```

The current PR is passing the tests, but the above versions of the functions are failing. Please let me know if you can spot the bug, I will look into it again when I find time.

**Comment by lucianolorenti** [comment]
Julio I saw your changes. The code looks great!
Regarding the last comment, the problem is that the conditions are being checked in the same image that is being modified. 
If the conditions are checked in `prev` and the update is done in `curr` the tests pass. I compared it with the matlab version in a few real images and gave me the same results.

```julia
function thinning_impl(img::AbstractArray{Bool,2}, algo::GuoAlgo)
    # pad input image
    prev = falses(size(img).+2)
    prev[2:end-1,2:end-1] = img

    # perform single iteration
    it = 1; curr = copy(prev)
    guo_iteration!(curr, prev, isodd(it))
    # loop if necessary
    while prev != curr
        it += 1; prev .= curr
        guo_iteration!(curr, prev, isodd(it))
    end

    # unpad result
    curr[2:end-1,2:end-1]
end

# perform Guo iteration on already padded image
function guo_iteration!(img::AbstractArray{Bool,2}, prev::AbstractArray{Bool,2}, odd_iteration::Bool)
    nx, ny = size(img)
    for j=2:ny-1, i=2:nx-1
        prev[i,j] || continue
        p1 = prev[i-1,j-1]
        p2 = prev[i-1,j]
        p3 = prev[i-1,j+1]
        p4 = prev[i,j+1]
        p5 = prev[i+1,j+1]
        p6 = prev[i+1,j]
        p7 = prev[i+1,j-1]
        p8 = prev[i,j-1]
        C = (!p2 && (p3 || p4)) + (!p4 && (p5 || p6)) + (!p6 && (p7 || p8)) + (!p8 && (p1 || p2))
        N1 = (p1 || p2) + (p3 || p4) + (p5 || p6) + (p7 || p8)
        N2 = (p2 || p3) + (p4 || p5) + (p6 || p7) + (p8 || p1)
        N = min(N1, N2)
        if odd_iteration
            O = (p2 || p3 || (!p5)) && p4
        else
            O = (p6 || p7 || (!p1)) && p8
        end
        if C == 1 && (2 ≤ N ≤ 3) && !O
            img[i,j] = false
        end
    end
end
```

**Comment by juliohm** [comment]
Hi @lucianolorenti sorry for the delay to reply, I am extremely busy around here. I thought the Guo iteration was supposed to take a single image as input and modify it in place, no?

When I wrote `prev .= curr` in the code, my intention was to copy the elements from the memory location `curr` to the memory location `prev` before overwriting `curr` in the next Guo iteration, is that not happening? I am still puzzled why the version I wrote is not producing the right result. Can you please clarify?

Basically, we should have two memory addresses `prev` and `curr`. Modify `curr` iteratively and save intermediate iterations in `prev`. Agree or I am missing something?

**Comment by juliohm** [comment]
Got it @lucianolorenti ! Thanks for clarifying. I will take a look at it again to see if the logic can be simplified and maybe rename the variables accordingly.

**Comment by juliohm** [comment]
Thanks again for clarifying @lucianolorenti , please let me know if you are fine with the changes, we should be ready to merge it after the tests pass :+1:

**Comment by juliohm** [comment]
Thank you very much @lucianolorenti for this great contribution! I will go ahead and release a new version of the package so that other people have access to it.

Feel free to optimize the implementation further in another PR, always welcome :+1:
